README.md
# Stmol
A component for building interactive molecular visualizations within Streamlit web-applications. It is designed to provide a state of the art visualizing and rendering 3D molecular structures for researchers, in a user-friendly manner.

------------------------------
## Stmol features 

- An easy-to-use component for rendering interactive 3D molecular visualizations of protein and ligand structures within Streamlit web-apps
- `stmol` can render protein and ligand structures with just a few lines of Python code by utilizing popular visualization libraries, currently Py3DMol and Speck. 
- On the user-end, `stmol` does not require expertise to interactively navigate. 
- On the developer-end, `stmol` can be easily integrated within structural bioinformatic and cheminformatic pipelines to provide a simple means for user-end researchers to advance biological studies and drug discovery efforts.
------------------------------

## Stmol installation

`stmol` runs with traditional Python stack.
To install `stmol` from [pypi](https://pypi.org/project/stmol/), run this command in your terminal:

``` console
pip install stmol==0.0.9
```

Since the end goal of `stmol` plugin, is to enable easy rendering of molecular structure within the streamlit applicaton, libraries such as 
[streamlit](https://github.com/streamlit/streamlit) and [py3Dmol](https://pypi.org/project/py3Dmol/) are crucial to work with. 
Run this command in your terminal to install the latest release of Streamlit,
``` console
pip install streamlit
```
Run this command in your terminal to install the latest release of py3Dmol,
``` console
pip install py3Dmol
```
A recent addition to `stmol` project, in a form of static HTML wrapper is the `speck_plot()` function, which helps in rendering [Speck](https://github.com/wwwtyro/speck) structures within the Streamlit web-application, is dependent on the following libraries, 
```console
ipyspeck(==0.6.1)
ipywidgets(==7.6.3)
```

-----------------

## Quickstart

##### 1. Protein visualization using `showmol()`function 

- **Using `py3Dmol object`**

To visualize any protein structure, all we need is the PDB ID of the protein.
```python
from stmol import showmol
import py3Dmol
# 1A2C
# Structure of thrombin inhibited by AERUGINOSIN298-A from a BLUE-GREEN ALGA
xyzview = py3Dmol.view(query='pdb:1A2C') 
xyzview.setStyle({'cartoon':{'color':'spectrum'}})
showmol(xyzview, height = 500,width=800)
```
- **Using the `render_pdb()` function**

The `render_pdb()` function accepts any PDB ID and returns a py3Dmol object. 
```python
from stmol import *
showmol(render_pdb(id = '1A2C'))
```
[![Quickstart-1](https://github.com/napoles-uach/stmol/blob/master/Resources/Quickstart-1.png)](https://github.com/napoles-uach/stmol/blob/master/Resources/Quickstart-1.png)

##### 2. Labelling protein using `render_pdb_resn()`function 
Inorder to mark the residues, we can use the `render_pdb_resn()` function, which in this example marks the *Alanine* [ALA] residues,
```python
showmol(render_pdb_resn(viewer = render_pdb(id = '1A2C'),resn_lst = ['ALA',]))
```
[![Quickstart-2](https://github.com/napoles-uach/stmol/blob/master/Resources/Quickstart-2.png)](https://github.com/napoles-uach/stmol/blob/master/Resources/Quickstart-2.png)

###### Refer to the documentation [here](https://napoles-uach-stmol-home-pom051.streamlitapp.com/Documentation) 
----------
## Examples
You can find several `stmol` examples [here](https://napoles-uach-stmol-home-pom051.streamlitapp.com/Examples).

### Browser support
- Chrome, Edge, Firefox - [link](https://napoles-uach.github.io/stmol/)
- Safari - [link](https://napoles-uach-stmol-home-pom051.streamlitapp.com)

----------------
## Dependencies
- [py3Dmol](https://pypi.org/project/py3Dmol/) 
- [ipyspeck](https://pypi.org/project/ipyspeck/)
- [ipywidgets](https://github.com/jupyter-widgets/ipywidgets)
- [streamlit](https://github.com/streamlit/streamlit)
-----------------

## Contribution to the project
We appreciate contributions from the community! Every little bit helps, and credit will always be given. 
- **Reporting Bugs** - Report bugs at https://github.com/napoles-uach/stmol/issues
- **Fix Bugs** - Look through the GitHub issues for bugs. Anything tagged with “bug” and “help wanted” is open to whoever wants to help with it.
- **Implement Features** - Look through the GitHub issues for features. Anything tagged with “enhancement” and “help wanted” is open to whoever wants to implement it.
- **Submit Feedback** - The best way to send feedback is to file an issue at https://github.com/napoles-uach/stmol/issues
--------------
## Authors
- J. M. Nápoles-Duarte <jnapoles@uach.mx>
- Avratanu Biswas <avratanu.biswas@brc.hu>
- Mitchell I. Parker <mip34@drexel.edu>


changelog.md
## Changelog

### 🚀 0.0.9 (Aug 12, 2022)

- :sparkles: Added features to work with `py3Dmol objects`
- :sparkles: Updated documentation page
- :bug: `speck_plot()` bug fixed

### 🚀 0.0.8 (Aug 12, 2022)

- :sparkles:`Stmol` webpage released 
- :tada: Documentation added 
- :bookmark: Pre-release of added features to work with `py3Dmol objects`

### 🚀 0.0.7 (Jan 6, 2022)

- :white_check_mark: `showmol()` well tested with demo apps  
- :zap: `Readme.md` updated 
- `component_3Dmol()` function deprecations :arrow_down:
- `speck_plot()` functionality adapted


### 🚀 0.0.5 (Aug 13, 2020)
- Pre-release :zap:

### 🚀 0.0.5 (Aug 2, 2020)
- Functionalities added :arrow_up:

### 🚀 0.0.5 (Aug 2, 2020)
- Testing and release

stmol/__init__.py
import os
import streamlit.components.v1 as components
import ipywidgets as widgets
from ipywidgets import embed
import py3Dmol
import ipyspeck

def showmol(mol_obj,height=500,width=500):
    """Shows the Py3DMOL object.

    Parameters
    ----------
    obj: Py3DMOL object
        Already existing Py3DMOL object, which can be created using the makeobj function.
    height: Integer, default 500
        Is the height of viwer window.
    width: Integer, default 500
        Is the width of wiewer window.
    
    Returns
    -------
    None.
    """
    components.html(mol_obj._make_html(), height = height,width=width)

def speck_plot(_xyz, wbox_height="700px", 
            wbox_width="800px",
            component_h = 700, 
            component_w = 800, 
            scroll = False):

    """ Plots the speckmol molecule using the ipyspeck library and returns 
        the <class 'ipyspeck.speck.Speck'>.
    
    Parameters
    ----------
    _xyz : str
        The xyz string of the molecule.
    wbox_height : str
        The height of the widget box.
    wbox_width : str
        The width of the widget box.
    component_h : int
        The height of the streamlit html component.
    component_w : int
        The width of the streamlit html component.
    scroll : bool
        If True, the streamlit component will scroll.  
    Returns
    -------
    spec_xyz : <class 'ipyspeck.speck.Speck'>
        The speckmol molecule. spec_xyz.keys() returns the keys of the
        molecule. For example - spec_xyz.keys() returns ['atomScale',  
        'bondScale', 'atomShade', 'bondThreshold', 'bondColor', 'atomColor',
        'outline', 'bonds', 'atomScale', 'atomColor', 'atomScale', 'atomScale]
        These keys are useful for modifying the molecule.
    """        
    # Read the xyz file
    spec_xyz = ipyspeck.speck.Speck(data = _xyz)
    # Create the widget box
    widg = widgets.Box([spec_xyz], layout=widgets.Layout(height=wbox_height,width=wbox_width))
    # Embed the widget box in the streamlit html component
    sc = embed.embed_snippet(widg)
    html = embed.html_template.format(title="", snippet=sc)
    components.html(html,height = component_h, width = component_w,scrolling=scroll)
    return spec_xyz

def makeobj(xyz,molformat='mol',style='stick',background='white'):
    """
    makeobj function accepts a molecule structure in a given format and returns a Py3DMOL molecule object.
    Accepted formats are 'mol', 'sdf', 'pdb', 'pqr', 'xyz'.

    Parameters
    ----------
    xyz: String
        Is the molecule object
    molformat: String, default 'mol'    
        Is the format of the molecule. Accepted formats are 'mol', 'sdf', 'pdb', 'pqr', 'xyz'.
    style: String, default 'stick'
        Is the style of the molecule. Can be 'stick', 'sphere', 'cross', 'surface', 'ribbon', 'cartoon'
    background: String, default 'white'
        Is the background color of the molecule
    
    Returns
    -------
    obj: Py3DMOL object
    """
    xyzview = py3Dmol.view()
    xyzview.addModel(xyz,molformat)
    xyzview.setStyle({style:{}})
    xyzview.setBackgroundColor(background)
    xyzview.zoomTo()
    return xyzview

def render_pdb(id='7T59'):
    """
    render_pdb function accepts a PDB ID and returns a Py3DMOL object.
    Example:
        obj=render_pdb()
        showmol(obj)

    Parameters
    ----------
    id: String, default '7T59'
        Is the PDB ID of the molecule.
    
    Returns
    -------
    obj: Py3DMOL object
    """
    viewer = py3Dmol.view(query=id)
    viewer.setStyle({ "cartoon": {
        "color": "spectrum",
        "colorReverse": True,
        "colorScale": "RdYlGn",
        "colorScheme": "Polarity",
        "colorBy": "resname",
            }})
    return viewer

def render_pdb_resn(viewer,resn_lst):
    """
    render_pdb_resn function accepts a Py3DMOL object and adds a list of PDB resn labels to it.
    Example:
        showmol(render_pdb_resn(render_pdb(),['ALA',]) )

    Parameters
    ----------
    viewer: Py3DMOL object
    resn_lst: List
        Is the list of PDB resn of the molecules. Example: ['ALA','ARG',,'LYS','THR','TRP','TYR','VAL']
    
    Returns
    -------
    obj: Py3DMOL object
    """
    viewer.setStyle({ "cartoon": {"color": "spectrum","colorBy": "resname",}})
    viewer.setStyle({'bonds':0},{'sphere':{'radius':0.3}})
    viewer.addResLabels({'resn':resn_lst,})
    return viewer

def render_pdb_resi(viewer,resi_lst):
    """
    render_pdb_resi function accepts a Py3DMOL object and adds a list of PDB resi labels to it.
    Example:
        showmol(render_pdb_resi(render_pdb(),['442-444']))

    Parameters
    ----------
    viewer: Py3DMOL object
    resi_lst: List
        Is the list of PDB resi of the molecules. Example: ['42-44','48','49']
    
    Returns
    -------
    obj: Py3DMOL object
    """
    viewer.setStyle({"cartoon": {"color": "spectrum","colorBy": "resname",}})
    viewer.setStyle({'bonds':0},{'sphere':{'radius':0.3}})
    viewer.addResLabels({'resi':resi_lst,})
    return viewer

def obj_upload(uploaded_file):
    """
    obj_upload function accepts a file uploaded with streamlit file_uploader function and returns a Py3DMOL object.
    Accepted formats are 'mol', 'sdf', 'pdb', 'pqr', 'xyz'.
    Example:
        uploaded_file = st.sidebar.file_uploader("Upload PDB file")
        if uploaded_file is not None:
            prot = obj_upload(uploaded_file)
            showmol(prot)

    Parameters
    ----------
    uploaded_file: File
        Is the PDB, xyz or any supported file of the molecule
    
    Returns
    -------
    obj: Py3DMOL object
    """
       
    pdbxyz = uploaded_file.getvalue().decode("utf-8")
    prot=makeobj(xyz=pdbxyz,molformat='pdb',style='stick',background='white')
    return prot

def add_model(obj,xyz,molformat='mol',model_style='stick'):
    """
    Adds a model to an existing Py3DMOL object.
    
    Parameters
    ----------
    obj: Py3DMOL object
        Already existing Py3DMOL object.
    xyz: String
        Is the model to be added.
    molformat: String, default 'mol'    
        Is the format of the added model
    model_style: String, default 'stick'
        Is the style of the added model. Can be 'stick', 'sphere', 'cross', 'surface', 'ribbon', 'cartoon'
    
    Returns
    -------
    None.
    """
    obj.addModel(xyz,molformat)
    obj.setStyle({'model':-1},{model_style:{}})

def add_hover(obj,backgroundColor='white',fontColor='black'):
    """
    Adds a hover function to the Py3DMOL object to show the atom name when the mouse hovers over an atom.
    Example:
        obj = render_pdb()
        add_hover(obj)
        showmol(obj)

    Parameters
    ----------
    obj: Py3DMOL object
        Already existing Py3DMOL object, which can be created using the makeobj function.
    backgroundColor: String, default 'white'
        Is the background color of the hover text
    fontColor: String, default 'black'
        Is the color of the text

    Returns
    -------
    None.
    """

    js_script = """function(atom,viewer) {
                   if(!atom.label) {
                    atom.label = viewer.addLabel(atom.atom+':'+atom.serial,{position: atom, backgroundColor:"%s" , fontColor:"%s"});
                }
              }"""%(backgroundColor,fontColor)
    obj.setHoverable({},True,js_script,
               """function(atom,viewer) {
                   if(atom.label) {
                    viewer.removeLabel(atom.label);
                    delete atom.label;
                   }
                }"""
               )

def add_box(obj,bxc=[0,0,0],bxd=[10,10,10],boxColor='blue',boxOpacity=0.5): 
    """
    Adds a box to an existing Py3DMOL object.
    Example:
        obj = render_pdb()
        add_box(obj)
        showmol(obj)

    Parameters
    ----------
    obj: Py3DMOL object 
        Already existing Py3DMOL object, which can be created using the makeobj function.
    bxc: List of Integers of len 3, default [0,0,0]
        Are the x,y,z coordinates of the center of the box in Angstroms
    bxd: List of Integers of len 3, default [10,10,10]
        Are the x,y,z dimensions of the box in Angstroms
    boxColor: String, default 'blue'
        Is the color of the box
    boxOpacity: String, default 0.5
        Is the opacity of the box
    
    Returns
    -------
    None.
    """
    obj.addBox({
        'center':{'x':bxc[0],'y':bxc[1],'z':bxc[2]},
        'dimensions': {'w':bxd[0],'h':bxd[1],'d':bxd[2]},
        'color':boxColor,
        'opacity': boxOpacity
        })

def add_sphere(obj,spcenter=[0,0,0],radius=10,spColor='blue'):
    """
    Adds a sphere to an existing Py3DMOL object.
    Example:
        obj = render_pdb()
        add_sphere(obj)
        showmol(obj)

    Parameters
    ----------
    obj: Py3DMOL object 
        Already existing Py3DMOL object, which can be created using the makeobj function.
    spcenter: List of Integers of len 3, default [0,0,0].
        Are the x,y,z coordinates of the center of the sphere in Angstroms
    radius: integer, default 10
        Is the radius of the sphere in Angstroms.
    spColor: String, default 'blue'
        Is the color of the sphere.
    
    Returns
    -------
    None.
    """
    obj.addSphere({
        'center':{'x':spcenter[0],'y':spcenter[1],'z':spcenter[2]},
        'radius':radius,
        'color':spColor
        })

def add_cylinder(obj,start=[0,0,0],end=[0,0,1],cylradius=1,cylColor='blue',dashed=True):
    """
    Adds a cylinder to an existing Py3DMOL object.
    Example:
        obj = render_pdb()
        add_cylinder(obj,end=[0,0,10])
        showmol(obj)

    Parameters
    ----------
    obj: Py3DMOL object
        Already existing Py3DMOL object, which can be created using the makeobj function.
    start: List of Integers of len 3, default [0,0,0]
        Are the x,y,z coordinates of the start of the cylinder in Angstroms.
    end: List of Integers of len 3, default [0,0,1]
        Are the x,y,z coordinates of the end of the cylinder in Angstroms.
    cylradius: integer, default 1
        Is the radius of the cylinder in Angstroms.
    cylColor: String, default 'blue'
        Is the color of the cylinder.
    dashed: Boolean, default True
        True for a cylinder dashed. False for a solid cylinder.
    
    Returns
    -------
    None.
    """

    obj.addCylinder({
        'start':{'x':start[0],'y':start[1],'z':start[2]},
        'end':{'x':end[0],'y':end[1],'z':end[2]},
        'radius':cylradius,
        'fromCap':True,
        'toCap':True,
        'color':cylColor,
        'dashed':dashed

        })
